### ISSUE TITLE
[🔷 Feature request ] Batch Back-Propogation

### ISSUE THREAD

**Issue by solid-droid** [matiasvlevi/Dann#41]
### Feature
Batch Back-Propagation / Training

### Type
- [x] Dann
- [ ] Matrix
- [ ] Layer
- [ ] Activation functions
- [ ] Loss functions
- [ ] Pool functions
- [ ] Datasets
- [ ] Documentation
- [ ] tests & examples
- [ ] Other 

### Description
In batch back-propagation, we split the full input set into smaller batches  
and  during training the model feed-forward through all the input / target pair of a batch 
without changing the weight / bias but will be updating the gradient (I am not sure on the math behind it).
once the batch is complete it updates the weights and biases based on a overall picture of batch rather than using one input.
this could help the model find the right weights and biases much faster than the single input back-propagation.

### Examples
How I would expect the final method to look like.  
```javascript
//say arch = 1 -> 2 - > 2

input = [1] 
target = [1 , 0]
nn.backpropagate(input , target)
// for one input, target pair

input =[ [ 1 ] , [-1 ] , [-5 ], [ 5 ] , [ 2 ] ]
target =[ [ 1 ,  0 ] , [ 0 , 1 ] , [ 0 , 1 ] , [ 1 , 0 ] , [ 1 , 0 ] ]

nn.backpropagate(input, target)
// expression remains same, internally it just need to check if Array.isArray(input[0]) => then batch train.
```
**Additional context**

I think this is available in all major ML libraries due to its great efficiency,
 also will help in creating distributed training capabilities.

**Comment by matiasvlevi** [comment]
would you like to be assigned to this issue?

**Comment by solid-droid** [comment]
Yeah, I can look into it. 
But it might take some time, as I really don't know the math behind it.

**Comment by matiasvlevi** [comment]
Perfect! If you ever need help, you can contact me on discord.

**Comment by solid-droid** [comment]
I was going through the source dann.js and got confused with the backpropagate method.  
Even though you have loss functions implemented, you are not using it to optimize the network ??
usually output layer activation is combined of loss and activation.
             like combination of softmax and mse 

error = target - predicted  :heavy_check_mark:
loss = lossFunc(error) :x:   //lossFunc = activation_and_lossFunc_combination
update w and b of output layer based on d/dx of loss :x:
update w and b based on d/dx of activation functions :heavy_check_mark:

https://github.com/matiasvlevi/Dann/blob/fdc7918f255d4efac3650245cb9afba97a856a56/build/dann.js#L2006-L2018
